Hokusai and Kubernetes

ashfurrow opened the pull request:
This post details deploying the new Apogee server to our Kubernetes cluster using Hokusai. I'm sure I got some of the technical nuances wrong here, so please comment where things are incorrect or could be further clarified.

acjay commented on _posts/2018-01-24-kubernetes-and-hokusai.markdown:
```diff
@@ -0,0 +1,134 @@
+
+Next up was to use Hokusai in our CI environment. [Circle CI 2.0][circle] is very Docker-oriented, so we set up everything using their [Workflows][]. This is a much higher level of abstraction for CI configuration than I'm used to, but I got the hang of it quickly. I created a job to run RSpec tests through Hokusai, a job to run [Danger][], a job to build and push a Docker image to our S3 bucket, ...
+
+Here's a complete, redacted copy of our Circle config:
```
Is "complete, redacted" contradictory? :) Maybe "redacted, but working"?

ashfurrow replied:
Mmm yeah.

ashfurrow commented:
@jonallured sadly, Apogee contains a lot of sensitive partner data surrounding how they format their data. We've decided to keep it closed source.

izakp commented on _posts/2018-01-24-kubernetes-and-hokusai.markdown:
```diff
@@ -0,0 +1,134 @@
+
+My background is in iOS software development, so spinning up new servers isn't something I do often. When I _do_, I usually use Heroku. After deploying to it, it feels like Kubernetes is a kind of hosted Heroku: it handles scaling up instances, managing worker/db/other instances, load-balancers, environment variables, promoting from staging to production – all that stuff. But Kubernetes' sophisti ...
+
+So basically, Kubernetes is to Heroku what Hokusai is to the Heroku command-line tool.
```
Hokusai was very much conceived with this intention!  I am very happy the point came across, without me actually stating it ^_^  However, maybe the wording order... "Hokusai is to Kubernetes what the Heroku CLI is to the Heroku platform"?

ashfurrow replied:
Agreed – I struggled with the wording there, I like that better.

izakp commented on _posts/2018-01-24-kubernetes-and-hokusai.markdown:
````diff
@@ -0,0 +1,134 @@
+            - push
+```
+
+The initial build on `master` built and pushed the server image, but the deploy failed. It's a Catch-22: we can't deploy until we have an image, but we only want to build images on CI, so we expect the first deploy to fail.
````
This is something we want to address, so feedback here would be helpful.  Maybe acknowledge Hokusai is still < v1.0 at this point?

ashfurrow replied:
Sure! I didn't find this to be a hindrance at all, really, but partially because I knew to expect it. I think probably the most straightforward solution might just be a helpful error message? Maybe after the failed deploy, Hokusai could check if the environment exists and provide a helpful explanation before failing? Or maybe it only prints the error message in a CI environment? In any case, I see this as more of a quirk than a bug. 

I didn't see an issue for it on Hokusai's repo, is there a public discussion you'd like linked to?

izakp replied:
Thanks - yes all part of the process.  I don't know the right solution tbh.  Tomorrow I'll make an issue for you to link to.  And maybe an apologetic line in that this is under development is appropriate is all.  I mean an apology [in the Platonic sense](https://en.wikipedia.org/wiki/Apology_(Plato)), as part of a discourse.

izakp replied:
https://github.com/artsy/hokusai/issues/50

ashfurrow commented:
@izakp I've addressed your feedback in https://github.com/artsy/artsy.github.io/pull/410/commits/78473cc3d3b3cccc5f8bed44fba2e9a1a49f8c6e – could you review and merge if everything looks good?
